### ISSUE TITLE
Problems with creating HashMap<&str, Box<dyn TraitName>> (need possibility to specify type manualy?)

### ISSUE THREAD

**Issue by Regela** [jofas/map_macro#2]
```
use map_macro::map;
use std::collections::HashMap;
use std::fmt;
use std::fmt::{Debug, Formatter};

struct S1 {}
struct S2 {}

trait Tr {}
impl Debug for dyn Tr {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        f.debug_struct("Tr").finish()
    }
}

impl Tr for S1 {}
impl Tr for S2 {}

fn foo(map: HashMap<&str, Box<dyn Tr>>) {
    println!("{:?}", map)
}

fn main() {
    // Compile error
    // foo(map! {
    //     "1" => Box::new(S1{}),
    //     "2" => Box::new(S2{}),
    // });

    // Compile error
    // foo({
    //     let map: HashMap<&str, Box<dyn Tr>> = map! {
    //             "1" => Box::new(S1{}),
    //             "2" => Box::new(S2{}),
    //     };
    //     map
    // });

    // Compile error
    // foo({
    //     let mut map: HashMap<&str, Box<dyn Tr>> = map! {
    //             "1" => Box::new(S1{}),
    //             "2" => Box::new(S2{}),
    //     };
    //     map
    // });

    // OK
    foo({
        let mut map: HashMap<&str, Box<dyn Tr>> = map! {};
        map.insert("1", Box::new(S1 {}));
        map.insert("2", Box::new(S2 {}));
        map
    })
}
```

# Compile error example
```
error[E0308]: mismatched types
   --> src/main.rs:27:25
    |
27  |         "2" => Box::new(S2{}),
    |                -------- ^^^^ expected struct `S1`, found struct `S2`
    |                |
    |                arguments to this function are incorrect
    |
note: associated function defined here
   --> /home/regela/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/boxed.rs:218:12
    |
218 |     pub fn new(x: T) -> Self {
    |            ^^^

error[E0308]: mismatched types
  --> src/main.rs:25:9
   |
25 |       foo(map! {
   |  _________^
26 | |         "1" => Box::new(S1{}),
27 | |         "2" => Box::new(S2{}),
28 | |     });
   | |_____^ expected trait object `dyn Tr`, found struct `S1`
   |
   = note: expected struct `HashMap<&str, Box<(dyn Tr + 'static)>>`
              found struct `HashMap<&str, Box<S1>>`
   = note: this error originates in the macro `map` (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to 2 previous errors
```

I need to create map with Box<dyn TraitName> inside. I cannot do it using your macros. How can this be done?

**Comment by jofas** [comment]
Hi @Regela,

unfortunately, this seems currently impossible, because the compiler is not able to figure out the correct type during type inference. Your call of the `map!` macro expands to something like:

```rust
foo({
  let mut map = HashMap::new();

  map.insert("1", Box::new(S1{}));
  map.insert("2", Box::new(S2{]));

  map
});
```

The compiler can't figure out the type of `map` properly, assuming it is `HashMap<&str, Box<S1>>` from the first call to `map.insert` and not -- as it is supposed to -- from the type that is expected from the surrounding context (in this case the `foo` function expecting a `HashMap<&str, Box<dyn Tr>>`).

This could be fixed by letting the compiler know which type we want the `map` variable in the macro to have, by providing the right type to the macro invocation. The best I came up with would be something like:

```rust
foo(map! {
  type=<&str, Box<dyn Tr>>,
  "1" => Box::new(S1{}),
  "2" => Box::new(S2{}),
});
```

This would work and I could add the optional `type=<...>` to the `map!` macro for you, but I'm not one hundred percent sure about the syntax. Do you have an idea or a preference concerning the syntax?

**Comment by Regela** [comment]
If we can't do something like 
```
map!<&str, Box<dyn Tr>>{
  "1" => Box::new(S1{}),
  "2" => Box::new(S2{}),
}
```
i think your variant good.

**Comment by jofas** [comment]
I found another solution to enable you to use `map-macro` with trait objects. One can use [type coercion](https://doc.rust-lang.org/reference/type-coercions.html) to let the compiler know what the right type is. I could add another macro `map_explicit!` (or something along that line) that uses explicit type coercion. 

```rust
foo(map_explicit! {
  "1" => Box::new(S1{}),
  "2" => Box::new(S2{}),
});
```
should work then. 

I think this is a better solution than adding `type=<...>` syntax to the existing `map!` macro

**Comment by jofas** [comment]
Just released `v0.2.5`. You can now use `map_macro::map_e` as a drop-in replacement for `map!`.

**Comment by Regela** [comment]
Yes, it worked for me, thank you! =)
